pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,1094 @@
+		k := f.Bytes()
+		if string(k) != key {
+			newThen = append(newThen, op)
+		}
+	}
+	txn.etcdThen = newThen
+
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+	return nil
+}
+
+func etcdDelRow(txn *Transaction, table, uuid string) error {
+	key := makeKey(txn.request.DBName, table, uuid)
+	etcdOp := clientv3.OpDelete(key)
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+	return nil
+}
+
+/* insert */
+func preInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName == "" {
+		return nil
+	}
+	etcdGetTable(txn, ovsOp.Table)
+	return nil
+}
+
+func doInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName != "" {
+		for _, row := range txn.data.Table(ovsOp.Table) {
+			if (*row)[COL_UUIDNAME] == ovsOp.UUIDName {
+				err := errors.New(E_DUP_UUIDNAME)
+				ovsResult.Error = err.Error()
+				return err
+			}
+		}
+	}
+	ovsResult.Count = ovsResult.Count + 1
+	uuid := guuid.New().String() /* generate RFC4122 UUID */
+	row := txn.data.Row(ovsOp.Table, uuid)
+	*row = ovsOp.Row // supporting read-afrer-write
+	return etcdPutRow(txn, ovsOp.Table, uuid, row)
+}
+
+/* select */
+func preSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return etcdGetByWhere(txn, ovsOp, ovsResult)
+}
+
+func doSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for _, row := range txn.data.Table(ovsOp.Table) {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		resultRow := reduceRowByColumns(row, ovsOp.Columns)
+		ovsResult.Count = ovsResult.Count + 1
+		ovsResult.Rows = append(ovsResult.Rows, *resultRow)
+	}
+	return nil
+}
+
+/* update */
+func preUpdate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return etcdGetByWhere(txn, ovsOp, ovsResult)
+}
+
+func doUpdate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for uuid, row := range txn.data.Table(ovsOp.Table) {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ovsResult.Count = ovsResult.Count + 1
+		err = RowUpdate(row, ovsOp.Row)
+		if err != nil {
+			return err
+		}
+		*(txn.data.Row(ovsOp.Table, uuid)) = *row
+		etcdPutRow(txn, ovsOp.Table, uuid, row)
+	}
+	return nil
+}
+
+/* mutate */
+func preMutate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return etcdGetByWhere(txn, ovsOp, ovsResult)
+}
+
+func doMutate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for uuid, row := range txn.data.Table(ovsOp.Table) {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ovsResult.Count = ovsResult.Count + 1
```

Comment:
result from etcd

Reply:
this result is as good as any. If the transaction was a success then this number of accurate

Reply:
>  If the transaction was a success then this number is accurate

It is not always true. for example delete row. Another point is a demonstration of which the operations failed.

Reply:
in addition, see my comment about the transaction response